### PR TITLE
Update _socket to use the socket2 crate; similar to the C sockets api

### PR DESCRIPTION
**Pull request by coolreader18**
Also adds `socket.py` from CPython 3.6.0

### CODE REVIEW

**Comment by palaviv** [comment]
@coolreader18 can you split this PR into smaller ones? It will be easier to review that way

**Comment by coolreader18** [comment]
@palaviv #1574

File: vm/src/stdlib/socket.rs [hunk 1]
```diff
@@ -1,493 +1,317 @@
...
-    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
-        match self {
-            Connection::TcpStream(con) => con.write(buf),
-            Connection::UdpSocket(con) => con.send(buf),
-            _ => Err(io::Error::new(io::ErrorKind::Other, "oh no!")),
-        }
-    }
-    fn flush(&mut self) -> io::Result<()> {
-        Ok(())
+        Ok(vm.ctx.new_tuple(vec![fd, addr_tuple]))
```
Comment (palaviv) [inline vm/src/stdlib/socket.rs]: You can return here `(fd, addr_tuple)`

File: vm/src/stdlib/socket.rs [hunk 2]
```diff
@@ -1,493 +1,317 @@
...
-                    }
-                } else {
-                    Err(vm.new_type_error("".to_string()))
-                }
+    #[pymethod]
+    fn recvfrom(&self, bufsize: usize, vm: &VirtualMachine) -> PyResult {
+        let mut buffer = vec![0u8; bufsize];
+        let addr = match self.sock().recv_from(&mut buffer) {
+            Ok((_, addr)) => addr,
+            Err(ref e) if e.kind() == io::ErrorKind::TimedOut => {
```
Comment (palaviv) [inline vm/src/stdlib/socket.rs]: I would suggest adding a function name `convert_socket_error` that will check if this is a timeout and then call `convert_io_error`. That will remove a lot of duplication.

**Comment by palaviv** [comment]
@coolreader18 can you look at #1580? The windows test of select failed there.

**Comment by coolreader18** [comment]
I'm not sure, that's a really weird failure. I tried rerunning it but the checkout failed because the branch is deleted.

**Pull request merged**